File: companies/Microsoft/3877.minimum-removals-to-achieve-target-xor.go
package leetcode

//
// @lc app=leetcode id=3877 lang=golang
//
// [3877] Minimum Removals to Achieve Target XOR
//
// Notes
// Keep as many elements as possible while obtaining each XOR value. dp[x] is the
// maximum kept count with XOR x after scanning a prefix; for every value we may
// skip it or keep it and move to x^value. The answer is n minus the best kept
// count for target. Time: O(n * 2^14). Space: O(2^14).
//
// @lc code=start

func MinRemovals3877(nums []int, target int) int {
	const maxXor = 1 << 14
	const unreachable = -1 << 30

	dp := make([]int, maxXor)
	for i := range dp {
		dp[i] = unreachable
	}
	dp[0] = 0

	for _, value := range nums {
		next := append([]int(nil), dp...)
		for xorValue, keptCount := range dp {
			if keptCount == unreachable {
				continue
			}
			newXor := xorValue ^ value
			if keptCount+1 > next[newXor] {
				next[newXor] = keptCount + 1
			}
		}
		dp = next
	}

	if dp[target] == unreachable {
		return -1
	}
	return len(nums) - dp[target]
}

// @lc code=end

File: pt-1/3877.minimum-removals-to-achieve-target-xor.java
/*
 * @lc app=leetcode id=3877 lang=java
 *
 * [3877] Minimum Removals to Achieve Target XOR
 *
 * Maximize how many elements we keep while obtaining each possible XOR.
 * dp[x] is the largest kept count producing XOR x. The answer removes all
 * other elements, or -1 if target is unreachable.
 *
 * Time: O(n * 2^14). Space: O(2^14).
 */

import java.util.Arrays;

// @lc code=start
class Solution {
    public int minRemovals(int[] nums, int target) {
        int maxXor = 1 << 14;
        int unreachable = -1_000_000_000;
        int[] dp = new int[maxXor];
        Arrays.fill(dp, unreachable);
        dp[0] = 0;

        for (int value : nums) {
            int[] next = dp.clone();
            for (int xor = 0; xor < maxXor; xor++) {
                if (dp[xor] == unreachable) {
                    continue;
                }
                int nx = xor ^ value;
                next[nx] = Math.max(next[nx], dp[xor] + 1);
            }
            dp = next;
        }

        return dp[target] == unreachable ? -1 : nums.length - dp[target];
    }
}
// @lc code=end
